File: 课程资料和情报（oj上的AC代码）/我是最快的马.cpp
//(1)没搞清楚憋马脚的含义（2）xx和yy是终点的话，不能标记vis
#include <bits/stdc++.h>
using namespace std;
int sx,sy,ex,ey,dx[8]={-1,1,-1,1,2,2,-2,-2},dy[8]={-2,-2,2,2,-1,1,-1,1},m,ok,num;
int ddx[8]={0,0,0,0,1,1,-1,-1},ddy[8]={-1,-1,1,1,0,0,0,0};
string anss;
set<pair<int,int>> dict,vis;
struct node{
    int x,y;
    string ans;
    node(){}
    node(int a,int b):x(a),y(b){
        ans="("+to_string(a)+","+to_string(b)+")";
    }
};
int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin>>sx>>sy>>ex>>ey>>m;
    for(int i=0;i<m;i++){
        int t1,t2;
        cin>>t1>>t2;
        dict.insert({t1,t2});
        vis.insert({t1,t2});
    }
    queue<node> q;
    q.push({sx,sy});
    while(!ok){
        int ssize=q.size();
        for(int ii=0;ii<ssize;ii++){
            node cur=q.front();
            if(cur.x==ex&&cur.y==ey){ok=1;anss=cur.ans;num++;}
            q.pop();
            for(int i=0;i<8;i++){
                int xx=cur.x+dx[i],yy=cur.y+dy[i];
                int xx1=cur.x+ddx[i],yy1=cur.y+ddy[i];
                if(vis.find({xx,yy})==vis.end()&&dict.find({xx1,yy1})==dict.end()){
                    if(xx!=ex||yy!=ey)vis.insert({xx,yy});
                    node tmp;
                    tmp.x=xx;tmp.y=yy;
                    tmp.ans=cur.ans+"-"+"("+to_string(xx)+","+to_string(yy)+")";
                    q.push(tmp);
                }
            }
        }
    }
    if(num==1)cout<<anss;
    else cout<<num;
    return 0;
}

File: 课程资料和情报（oj上的AC代码）/我是最快的马.py
from collections import deque
sx,sy=map(int,input().split())
ex,ey=map(int,input().split())
p=set([tuple(map(int,input().split()))for _ in range(int(input()))])
q,dx,dy,vis=deque([[sx,sy,'('+str(sx)+','+str(sy)+')']]),[-2,-2,-1,1,2,2,1,-1],[1,-1,-2,-2,-1,1,2,2],set([(sx,sy)])
ans,cur=0,''
while q:
    tmp=deque()
    while q:
        x,y,path=q.popleft()
        wx,wy=[-1,0,1,0],[0,-1,0,1]
        if x==ex and y==ey:
            ans+=1
            if ans==1:cur=path
        for i in range(8):
            xx,yy,x1,y1=x+dx[i],y+dy[i],x+wx[i//2],y+wy[i//2]
            if (xx,yy)not in vis and (x1,y1)not in p:tmp.append([xx,yy,path+'-'+'('+str(xx)+','+str(yy)+')'])
    if ans:break
    q=tmp
print(cur if ans==1 else ans)